lib/Arch.cpp

```diff
@@ -0,0 +1,441 @@
+#include <vector>
+
+#include "anvill/Arch.h"
+#include "anvill/Decl.h"
+
+#include <glog/logging.h>
+#include <remill/Arch/Arch.h>
+
+namespace remill {
+class Arch;
+class IntrinsicTable;
+struct Register;
+}  // namespace remill
+
+namespace anvill {
+
+// Try to recover parameter names using debug information. Otherwise, name the
+// parameters with the form "param_x". The mapping of the return value is
+// positional starting at 1.
+std::map<unsigned, std::string> TryRecoverParamNames(
+    const llvm::Function &function) {
+  std::map<unsigned int, std::string> param_names;
+
+  for (auto &block : function) {
+    for (auto &inst : block) {
+      if (auto debug_inst = llvm::dyn_cast<llvm::DbgInfoIntrinsic>(&inst)) {
+        if (auto value_intrin = llvm::dyn_cast<llvm::DbgDeclareInst>(&inst)) {
+          const llvm::MDNode *mdn = value_intrin->getVariable();
+          const llvm::DILocalVariable *div =
+              llvm::cast<llvm::DILocalVariable>(mdn);
+
+          // Make sure it is actually an argument
+          if (div->getArg() != 0) {
+            LOG(INFO) << div->getArg() << " : " << div->getName().data();
+            param_names[div->getArg()] = div->getName().data();
+          }
+        } else if (auto value_intrin =
+                       llvm::dyn_cast<llvm::DbgValueInst>(debug_inst)) {
+          const llvm::MDNode *mdn = value_intrin->getVariable();
+          const llvm::DILocalVariable *div =
+              llvm::cast<llvm::DILocalVariable>(mdn);
+
+          if (div->getArg() != 0) {
+            LOG(INFO) << div->getArg() << " : " << div->getName().data();
+            param_names[div->getArg()] = div->getName().data();
+          }
+        }
+      }
+    }
+  }
+
+  // If we don't have names for some parameters then automatically name them
+  unsigned int num_args =
+      (unsigned int)(function.args().end() - function.args().begin());
+  for (unsigned int i = 1; i <= num_args; i++) {
+    if (!param_names.count(i)) {
+      param_names[i] = "param" + std::to_string(i);
+    }
+  }
+
+  return param_names;
+}
+
+// TODO: This doesn't really fit here so I will need to find a better place for
+// this. Translates an llvm::Type to a type that conforms to the spec in
+// TypeParser.cpp
+std::string TranslateType(const llvm::Type &type) {
```

Comment:
Instead of returning a `std::string`, I recommend having this function take in a `std::stringstream` by reference and inserting into that. Also recommend taking in a data layout and a unordered map of struct type pointers to integer IDs.

If you want an interface that returns a std::string, then have that as a second function, wrapping this one, where it takes a module pointer and a type pointer, returns a struct, and calls this function.

Reply:
Maybe make a `TypePrinter.cpp` or `TypeUnparser.cpp` or `TypeSerializer.cpp`